lib/widgets/compose_box.dart

```diff
@@ -234,61 +307,110 @@ class _AttachFileButton extends StatelessWidget {
+      assert(f.readStream != null);  // We passed `withReadStream: true` to pickFiles.
+      return _File(content: f.readStream!, length: f.size, filename: f.name);
+    });
+    _uploadFiles(context: context, contentController: contentController, contentFocusNode: contentFocusNode,
+      files: files);
+  }
 
-    final List<PlatformFile> tooLargeFiles = [];
-    final List<PlatformFile> rightSizeFiles = [];
-    for (PlatformFile file in result.files) {
-      if ((file.size / (1 << 20)) > store.maxFileUploadSizeMib) {
-        tooLargeFiles.add(file);
-      } else {
-        rightSizeFiles.add(file);
-      }
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(icon: const Icon(Icons.attach_file), onPressed: () => _handlePress(context));
+  }
+}
+
+class _AttachMediaButton extends StatelessWidget {
+  const _AttachMediaButton({required this.contentController, required this.contentFocusNode});
+
+  final ContentTextEditingController contentController;
+  final FocusNode contentFocusNode;
+
+  _handlePress(BuildContext context) async {
+    FilePickerResult? result;
+    try {
+      result = await FilePicker.platform.pickFiles(type: FileType.media, allowMultiple: true, withReadStream: true);
+    } catch (e) {
+      // TODO(i18n)
+      showErrorDialog(context: context, title: 'Error', message: e.toString());
+      return;
+    }
+    if (result == null) {
+      return; // User cancelled; do nothing
     }
 
-    if (tooLargeFiles.isNotEmpty) {
-      final listMessage = tooLargeFiles
-        .map((file) => '${file.name}: ${(file.size / (1 << 20)).toStringAsFixed(1)} MiB')
-        .join('\n');
-      showErrorDialog( // TODO(i18n)
-        context: context,
-        title: 'File(s) too large',
-        message:
-          '${tooLargeFiles.length} file(s) are larger than the server\'s limit of ${store.maxFileUploadSizeMib} MiB and will not be uploaded:\n\n$listMessage');
+    if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
+    else {
+      return;
     }
 
-    final List<(int, PlatformFile)> uploadsInProgress = [];
-    for (final file in rightSizeFiles) {
-      final tag = contentController.registerUploadStart(file.name);
-      uploadsInProgress.add((tag, file));
+    final Iterable<_File> files = result.files.map((f) {
+      assert(f.readStream != null);  // We passed `withReadStream: true` to pickFiles.
+      return _File(content: f.readStream!, length: f.size, filename: f.name);
+    });
+    _uploadFiles(context: context, contentController: contentController, contentFocusNode: contentFocusNode,
+        files: files);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(icon: const Icon(Icons.image), onPressed: () => _handlePress(context));
+  }
+}
+
+class _AttachFromCameraButton extends StatelessWidget {
+  _AttachFromCameraButton({required this.contentController, required this.contentFocusNode});
+
+  final _picker = ImagePicker();
```

Comment:
It makes me a bit nervous to be constructing this in an initializer on a widget.

I think the right fix is that we should just construct it in `_handlePress`. If there were some way that could go wrong on account of constructing it multiple times, then we'd be exposed to the same problem from constructing it here — because we'll be constructing new `_AttachFromCameraButton` widgets each time `StreamComposeBox` rebuilds.

In fact, from the implementation, it looks like `ImagePicker` has no non-static properties at all — its methods (except for `==` and `hashCode` inherited from `Object`) all behave exactly the same on any instance, as if they were static methods. That means that where and when we construct it really doesn't matter.

So probably the methods *should* be static methods; or the class should have a const constructor, which would accomplish much the same thing (and be a smoother migration from the status quo). But because they're not and it doesn't, the API isn't promising that we have that flexibility. I'd therefore like to avoid the pattern that looks like a bug and makes me nervous :slightly_smiling_face: